### PR TITLE
BAC-62: SDK for Clarify API v1.1beta2

### PR DESCRIPTION
**Pull request by smyrman**
commit effe8b03f46af3823667fee72c223cc77dbcde86 (HEAD -> work, origin/work)
Author: Sindre Myren <sindre@clarify.io>
Date:   Mon May 30 11:30:20 2022 +0200

    jsonrpc: include more information in decode errors
    
    Include more information in cases where we fail to decode the HTTP
    response into an RPC response, including traceparent header and parsed
    portions of the response body.

commit ff36dac8f2c0bc6bc0bad50885932b2dfabbfceb
Author: Sindre Myren <sindre@clarify.io>
Date:   Thu Mar 24 13:16:32 2022 +0100

    breaking: refactor and switch to API 1.1beta2
    
    There are several breaking changes, including a complete restructure of
    the package layout:
    
    - jsonrpc/resource: reusable models for requests and responses.
    - fields: reusable data and meta-data field types.
    - views: data and meta-data resource views.
    
    Fixups:
    
    - Add missing data parameter for savesignal
    - Add missing integration for insertRequest
    - Add missing 0 to make() for params
    
    Add support and examples for Clarify API v1.1beta2 methods:
    
    - admin.selectSignals
    - admin.publishSignals
    - clarify.selectItems
    - clarify.dataFrame

commit 64097d9985581d07cebf38a0507d16873331c768
Author: Sindre Myren <sindre@clarify.io>
Date:   Thu Apr 7 20:43:38 2022 +0200

    editor: add .cspell.json file
    
    Editors that are configured to use cspell will discover this file and
    use it, leading to fewer false positives for non-existing words.

commit d08aa673a53c069c1fc0ee0cbd3df86fb943a3d8
Author: Sindre Myren <sindre@clarify.io>
Date:   Tue Apr 5 22:43:46 2022 +0200

    ci: add go-header check to golangci-lint

commit 89ad69fe9582944e0ec1689ab958dd3a2e7b4305
Author: Sindre Myren <sindre@clarify.io>
Date:   Tue Apr 5 22:43:14 2022 +0200

    .gitignore: add VS Code config

### CODE REVIEW

File: jsonrpc/resource/resource.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	Data []Identifier               `json:"data"`
+}
+
+// Identifier uniquely identifies a resource entry.
+type Identifier struct {
+	Type string `json:"type"`
+	ID   string `json:"id"`
+}
+
+// NullIdentifier is a version of Identifier where the zero-value is encode as
```
Comment (sarzhann) [inline jsonrpc/resource/resource.go:98]: nit: `zero-value is encoded`

File: jsonrpc/resource/response.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resource
+
+// SaveSummary holds the identity of a resource, and describe weather it was
```
Comment (sarzhann) [inline jsonrpc/resource/response.go:17]: nit: `describes`

File: jsonrpc/resource/response.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+// updated or created due to your operation.
+type SaveSummary struct {
+	ID      string `json:"id"`
+	Created bool   `json:"created"`
+	Updated bool   `json:"updated"`
+}
+
+// CreateSummary holds the identity of a resource, and describe weather it was
```
Comment (sarzhann) [inline jsonrpc/resource/response.go:25]: nit: `describes`

File: query/comparison.go [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+	}
+
+	var target struct {
+		opComparison
+		NotEqual json.RawMessage `json:"$ne,omitempty"`
+	}
+
+	switch data[0] {
+	case '{':
+		// got object, treat as operator comparison.
```
Comment (sarzhann) [inline query/comparison.go:257]: nit: `Got object`

**Review by sarzhann — COMMENTED** [review]
Overall look good for me, left small typos

File: examples/publish_signals/main.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	creds, err := clarify.CredentialsFromFile("clarify-credentials.json")
+	if err != nil {
+		panic(err)
+	}
+
+	ctx := context.Background()
+	client := creds.Client(ctx)
+
+	// To select or publish signals, you must enable access to the "admin"
+	// namespace in clarify; this allows selecting and publishing signals from all integrations
```
Comment (sarzhann) [inline examples/publish_signals/main.go:37]: In my case it works only when both `admin` and `clarify` namespaces enabled. Is it intended by the docstring?

File: examples/select_items/main.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+	}
+
+	ctx := context.Background()
+	client := creds.Client(ctx)
+
+	// To select or publish signals, you must enable access to the "admin"
+	// namespace in clarify; this allows selecting signals from all integrations
+	// in the organization, but for the example, we will use the integration
+	// from the credentials file.
+	result, err := client.SelectItems().Filter(query.Comparisons{
```
Comment (sarzhann) [inline examples/select_items/main.go:25]: I get `Method not found` for `clarify.SelectItems` request

File: examples/publish_signals/main.go (hunk 5 above)
Comment (smyrman) [reply examples/publish_signals/main.go:37]: It's not intended; might be a bug in the API. I will double check it.

File: examples/select_items/main.go (hunk 6 above)
Comment (smyrman) [reply examples/select_items/main.go:25]: Which credentials are you using? Should work against dev (clarifyapp.clarify.cloud) and (clarifyapp.clarify.io) instances, but won't work against the scratch1 instance, unless you have rebased on master.

Comment (sarzhann) [reply examples/select_items/main.go:25]: I've tested on `dev` and all methods work fine except of `SelectItems`, but I'll check it again

Comment (sarzhann) [reply examples/select_items/main.go:25]: Tested on `dev` and `prod` and now it works fine. Probably forgot to update credentials for this case

Comment (smyrman) [reply examples/select_items/main.go:25]: Thanks for re-testing 🙇

File: examples/publish_signals/main.go (hunk 5 above)
Comment (smyrman) [reply examples/publish_signals/main.go:37]: I found the bug in the API; let's close this issue from an SDK point of view; the fix needs to go into the API.

**Review by sarzhann — APPROVED** [review]

**Pull request merged**